### PR TITLE
TS-3478: Indexing header representations on HPACK encoder

### PR DESCRIPTION
**Pull request by syucream**
https://issues.apache.org/jira/browse/TS-3478

### CODE REVIEW

File: proxy/http2/HPACK.h [hunk 1]
```diff
@@ -153,9 +165,9 @@ int64_t decode_string(Arena &arena, char **str, uint32_t &str_length, const uint
 
 int64_t encode_indexed_header_field(uint8_t *buf_start, const uint8_t *buf_end, uint32_t index);
 int64_t encode_literal_header_field(uint8_t *buf_start, const uint8_t *buf_end, const MIMEFieldWrapper &header, uint32_t index,
-                                    HpackFieldType type);
+                                    Http2DynamicTable &dynamic_table, HpackFieldType type);
 int64_t encode_literal_header_field(uint8_t *buf_start, const uint8_t *buf_end, const MIMEFieldWrapper &header,
```
Comment (masaori335) [inline proxy/http2/HPACK.h]: Those `encode_literal_header_field()` should be something like `encode_literal_header_field_with_indexed_name()` and `encode_literal_header_field_with_new_name()` ?

File: proxy/http2/HPACK.cc [hunk 2]
```diff
@@ -263,12 +263,80 @@ Http2DynamicTable::add_header_field(const MIMEField *field)
     }
 
     MIMEField *new_field = _mhdr->field_create(name, name_len);
+    new_field->name_set(_mhdr->m_heap, _mhdr->m_mime, name, name_len);
```
Comment (maskit) [inline proxy/http2/HPACK.cc]: It seems the field name has been set just one line above. Why is this needed?

File: proxy/http2/HPACK.cc [hunk 3]
```diff
@@ -263,12 +263,80 @@ Http2DynamicTable::add_header_field(const MIMEField *field)
...
     new_field->value_set(_mhdr->m_heap, _mhdr->m_mime, value, value_len);
+    mime_hdr_field_attach(_mhdr->m_mime, new_field, 1, NULL);
```
Comment (maskit) [inline proxy/http2/HPACK.cc]: Why don't you use `MIMEHdr::field_attach`? It's inconsistent to use the function here.

File: proxy/http2/HPACK.cc [hunk 4]
```diff
@@ -263,12 +263,80 @@ Http2DynamicTable::add_header_field(const MIMEField *field)
...
     // XXX Because entire Vec instance is copied, Its too expensive!
     _headers.insert(0, new_field);
   }
 }
 
+uint32_t
+Http2DynamicTable::get_dynamic_table_size() const
```
Comment (maskit) [inline proxy/http2/HPACK.cc]: I prefer `get_table_size` because we know it is Dynamic Table.

File: proxy/http2/HPACK.cc [hunk 5]
```diff
@@ -263,12 +263,80 @@ Http2DynamicTable::add_header_field(const MIMEField *field)
...
+  const char *target_value = field.value_get(&target_value_len);
+  const int entry_num = TS_HPACK_STATIC_TABLE_ENTRY_NUM + get_current_entry_num();
+
+  for (int index = 1; index < entry_num; ++index) {
+    const char *table_name, *table_value;
+    int table_name_len = 0, table_value_len = 0;
+
+    if (index < TS_HPACK_STATIC_TABLE_ENTRY_NUM) {
+      // static table
+      table_name = STATIC_TABLE[index].name;
```
Comment (maskit) [inline proxy/http2/HPACK.cc]: Should we care about Static Table here? It's Http2DynamicTable class.

File: proxy/http2/HPACK.cc [hunk 6]
```diff
@@ -263,12 +263,80 @@ Http2DynamicTable::add_header_field(const MIMEField *field)
...
+    } else {
+      // dynamic table
+      const MIMEField *m_field = get_header(index - TS_HPACK_STATIC_TABLE_ENTRY_NUM + 1);
+      table_name = m_field->name_get(&table_name_len);
+      table_value = m_field->value_get(&table_value_len);
+    }
+
+    // Check whether name (and value) are matched
+    if (ptr_len_casecmp(target_name, target_name_len, table_name, table_name_len) == 0) {
+      if (ptr_len_casecmp(target_value, target_value_len, table_value, table_value_len) == 0) {
```
Comment (maskit) [inline proxy/http2/HPACK.cc]: At least values must be compared with case sensitive.

**Comment by maskit** [comment]
Http2DynamicTable class is getting messy. To clarify the responsibility of Http2DynamicTable and to make the codes maintainable, it would be better to cleanup at this time.

I suggest having Indexing Tables layer and accessors for Static Table instead of adding all the things into Http2DynamicTable. It would make the codes much readable.

File: proxy/http2/HPACK.h (hunk 1 above)
Comment (syucream) [inline proxy/http2/HPACK.h]: I prefer `with_(indexed|new)_name`. I'll rename them.

**Comment by syucream** [comment]
@maskit I concern about Http2DynamicTable too. I'll add IndexingTable layer and separate Http2DynamicTable and static table.

**Comment by syucream** [comment]
@masaori335 @maskit please take another look

**Comment by masaori335** [comment]
I'm fine to ship this.

1. Actually I'm not big fun of names of classes and functions in HPACK starting with 'HTTP2'. But those confusions are going to be fixed in [TS-3967](https://issues.apache.org/jira/browse/TS-3967) with others.
2. Regression tests worked.

**Comment by maskit** [comment]
The fixes look ok to me.

@masaori335 You mean TS-4092? I'm going to refine the names and interfaces.

**Comment by masaori335** [comment]
Oops, I mean TS-4092.
